internal/sdkprovider/service/kafka/kafka_native_acl.go

```diff
@@ -0,0 +1,169 @@
+	"resource_type": {
+		Type:         schema.TypeString,
+		Required:     true,
+		ForceNew:     true,
+		ValidateFunc: validation.StringInSlice(kafka.ResourceTypeChoices(), false),
+		Description:  userconfig.Desc("The kafka resource type").ForceNew().PossibleValuesString(kafka.ResourceTypeChoices()...).Build(),
+	},
+	"pattern_type": {
+		Type:         schema.TypeString,
+		Required:     true,
+		ForceNew:     true,
+		ValidateFunc: validation.StringInSlice(kafka.PatternTypeChoices(), false),
+		Description:  userconfig.Desc("Resource pattern used to match specified resources").ForceNew().PossibleValuesString(kafka.PatternTypeChoices()...).Build(),
+	},
+	"principal": {
+		Type:         schema.TypeString,
+		Required:     true,
+		ForceNew:     true,
+		ValidateFunc: validation.StringLenBetween(1, 256),
+		Description:  userconfig.Desc("Principal is in type:name' format").ForceNew().MaxLen(256).Build(),
+	},
+	"host": {
+		Type:         schema.TypeString,
+		Optional:     true,
+		ForceNew:     true,
+		ValidateFunc: validation.StringLenBetween(1, 256),
+		Description:  userconfig.Desc("The host or `*` for all hosts").ForceNew().MaxLen(256).Build(),
+	},
+	"operation": {
+		Type:         schema.TypeString,
+		Required:     true,
+		ForceNew:     true,
+		ValidateFunc: validation.StringInSlice(kafka.OperationTypeChoices(), false),
+		Description:  userconfig.Desc("The operation").ForceNew().PossibleValuesString(kafka.OperationTypeChoices()...).Build(),
+	},
+	"permission_type": {
+		Type:         schema.TypeString,
+		Required:     true,
+		ForceNew:     true,
+		ValidateFunc: validation.StringInSlice(kafka.KafkaAclPermissionTypeChoices(), false),
+		Description:  userconfig.Desc("The permission type").ForceNew().PossibleValuesString(kafka.KafkaAclPermissionTypeChoices()...).Build(),
+	},
+}
+
+func resourceKafkaNativeACLCreate(ctx context.Context, d *schema.ResourceData, client avngen.Client) error {
+	var req kafka.ServiceKafkaNativeAclAddIn
+	err := schemautil.ResourceDataGet(d, &req)
+	if err != nil {
+		return err
+	}
+
+	project := d.Get("project").(string)
+	serviceName := d.Get("service_name").(string)
+
+	acl, err := client.ServiceKafkaNativeAclAdd(
+		ctx,
+		project,
+		serviceName,
+		&req,
+	)
+	if err != nil {
+		return err
+	}
+
+	err = schemautil.ResourceDataSet(aivenKafkaNativeACLSchema, d, acl)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(schemautil.BuildResourceID(project, serviceName, acl.Id))
+	return resourceKafkaNativeACLRead(ctx, d, client)
+}
+
+func resourceKafkaNativeACLRead(ctx context.Context, d *schema.ResourceData, client avngen.Client) error {
+	project, serviceName, aclID, err := schemautil.SplitResourceID3(d.Id())
+	if err != nil {
+		return err
+	}
+
+	acl, err := client.ServiceKafkaNativeAclGet(
+		ctx,
+		project,
+		serviceName,
+		aclID,
+	)
+	if err != nil {
+		return err
+	}
+
+	if acl == nil {
+		return fmt.Errorf(
+			errmsg.AivenResourceNotFound,
+			"aiven_kafka_native_acl",
+			aclID,
+		)
+	}
```

Comment:
1. `errmsg.AivenResourceNotFound` comes from the plugin framework, and it might produce cycled imports later. Please use just `fmt.Error("foo %w:", err)`
2. According the convention, you either get `err` or `acl` and never both or one. Error check is sufficient, 122-128 can be removed